nat/behavior/discovery.go

```diff
@@ -0,0 +1,365 @@
+	}
+	if myIP == outboundIP.String() {
+		return MAPPING_NONE, nil
+	}
+
+	// Test II: Send binding request to the other address but primary port
+	ctx1, cl = context.WithTimeout(ctx, timeout)
+	defer cl()
+	oaddr := *mapTestConn.OtherAddr
+	oaddr.Port = mapTestConn.RemoteAddr.Port
+	resp, err = mapTestConn.roundTrip(ctx1, request, &oaddr)
+	if err != nil {
+		return "", errors.Wrap(err, "mapping test II RT failed")
+	}
+
+	// Assert mapping behavior
+	resps2 := parse(resp)
+	if resps2.xorAddr.String() == resps1.xorAddr.String() {
+		return MAPPING_INDEPENDENT, nil
+	}
+
+	// Test III: Send binding request to the other address and port
+	ctx1, cl = context.WithTimeout(ctx, timeout)
+	defer cl()
+	resp, err = mapTestConn.roundTrip(ctx1, request, mapTestConn.OtherAddr)
+	if err != nil {
+		return "", errors.Wrap(err, "mapping test III RT failed")
+	}
+
+	// Assert mapping behavior
+	resps3 := parse(resp)
+	if resps3.xorAddr.String() == resps2.xorAddr.String() {
+		return MAPPING_ADDRESS_DEPENDENT, nil
+	}
+
+	return MAPPING_ADDRESSPORT_DEPENDENT, nil
+}
+
+// DiscoverNATFiltering returns either one of FILTERING_* constants describing
+// NAT filtering behavior or error
+func DiscoverNATFiltering(ctx context.Context, addrStr string, timeout time.Duration) (string, error) {
+	mapTestConn, err := connect(addrStr)
+	defer mapTestConn.Close()
+	if err != nil {
+		return "", errors.Wrap(err, "STUN connection init failed")
+	}
+
+	// Test I: Regular binding request
+	request := stun.MustBuild(stun.TransactionID, stun.BindingRequest)
+
+	ctx1, cl := context.WithTimeout(ctx, timeout)
+	defer cl()
+	resp, err := mapTestConn.roundTrip(ctx1, request, mapTestConn.RemoteAddr)
+	if err != nil {
+		return "", errors.Wrap(err, "filtering test I RT failed")
+	}
+	resps := parse(resp)
+	if resps.xorAddr == nil {
+		return "", errors.Wrap(errNoXorAddress, "filtering test I got bad response")
+	}
+
+	if resps.otherAddr == nil {
+		return "", errors.Wrap(errNoOtherAddress, "filtering test I got bad response")
+	}
+	addr, err := net.ResolveUDPAddr("udp4", resps.otherAddr.String())
+	if err != nil {
+		return "", errors.Wrap(err, "other-address resolve failed")
+	}
+	mapTestConn.OtherAddr = addr
+
+	// Test II: Request to change both IP and port
+	request = stun.MustBuild(stun.TransactionID, stun.BindingRequest)
+	request.Add(stun.AttrChangeRequest, []byte{0x00, 0x00, 0x00, 0x06})
+
+	ctx1, cl = context.WithTimeout(ctx, timeout)
+	defer cl()
+	resp, err = mapTestConn.roundTrip(ctx1, request, mapTestConn.RemoteAddr)
+	switch {
+	case err == nil:
+		return FILTERING_INDEPENDENT, nil
+	case err == ctx1.Err() && ctx.Err() == nil:
+		// Nothing, just no response. Proceed to next test.
+	default:
+		return "", errors.Wrap(err, "filtering test II failed")
+	}
+
+	// Test III: Request to change port only
+	request = stun.MustBuild(stun.TransactionID, stun.BindingRequest)
+	request.Add(stun.AttrChangeRequest, []byte{0x00, 0x00, 0x00, 0x02})
```

Comment:
I am surprised that `stun` package does not have predefined constants for `[]byte{0x00, 0x00, 0x00, 0x02}` or `[]byte{0x00, 0x00, 0x00, 0x06}` or `[]byte{0x00, 0x00, 0x00, 0x04}`

Maybe it makes more sense to have these as variables with meaningful names `var changePort = []byte{0x00, 0x00, 0x00, 0x02}` etc... instead of brittle comments that developers tend to forget to update ?

Reply:
Moved to named constants